diff -u a/platform.c b/platform.c
--- a/platform.c	2015-06-30 21:35:31.000000000 -0500
+++ b/platform.c	2017-01-30 11:14:09.914618209 -0600
@@ -112,6 +112,8 @@
 		solaris_set_default_project(pw);
 #endif
 
+#ifndef __PASE__
+
 #if defined(HAVE_LOGIN_CAP) && defined (__bsdi__)
 	if (getuid() == 0 || geteuid() == 0)
 		setpgid(0, 0);
@@ -136,6 +138,8 @@
 			error("setluid: %s", strerror(errno));
 	}
 #endif
+
+#endif
 }
 
 /*
@@ -145,6 +149,8 @@
 void
 platform_setusercontext_post_groups(struct passwd *pw)
 {
+#ifndef __PASE__
+
 #if !defined(HAVE_LOGIN_CAP) && defined(USE_PAM)
 	/*
 	 * PAM credentials may take the form of supplementary groups.
@@ -186,6 +192,8 @@
 #ifdef WITH_SELINUX
 	ssh_selinux_setup_exec_context(pw->pw_name);
 #endif
+
+#endif
 }
 
 char *
@@ -205,6 +211,22 @@
 int
 platform_sys_dir_uid(uid_t uid)
 {
+#ifdef __PASE__
+       static uid_t qsys_uid = -1;
+
+       if(qsys_uid == (uid_t)-1)
+       {
+               struct passwd* pw;
+               struct passwd pw_struct;
+               char pw_buf[1024];
+               getpwnam_r("qsys", &pw_struct, pw_buf, sizeof(pw_buf), &pw);
+               if(pw)
+                       qsys_uid = pw->pw_uid;
+       }
+
+       if (uid == qsys_uid)
+                       return 1;
+#endif
        if (uid == 0)
                return 1;
 #ifdef PLATFORM_SYS_DIR_UID

