--- a/servconf.c	2015-06-30 21:35:31.000000000 -0500
+++ b/servconf.c	2017-01-30 11:19:56.561664463 -0600
@@ -167,6 +167,9 @@
 	options->ip_qos_bulk = -1;
 	options->version_addendum = NULL;
 	options->fingerprint_hash = -1;
+#ifdef __PASE__
+	options->pase_shell = NULL;
+#endif
 }
 
 /* Returns 1 if a string option is unset or set to "none" or 0 otherwise. */
@@ -329,6 +332,10 @@
 	}
 	if (options->permit_tun == -1)
 		options->permit_tun = SSH_TUNMODE_NO;
+#ifdef __PASE__
+	if (options->pase_shell == NULL)
+		options->pase_shell = xstrdup(_PATH_BSHELL);
+#endif
 	if (options->ip_qos_interactive == -1)
 		options->ip_qos_interactive = IPTOS_LOWDELAY;
 	if (options->ip_qos_bulk == -1)
@@ -412,6 +419,9 @@
 	sAuthenticationMethods, sHostKeyAgent, sPermitUserRC,
 	sStreamLocalBindMask, sStreamLocalBindUnlink,
 	sAllowStreamLocalForwarding, sFingerprintHash,
+#ifdef __PASE__
+	sIBMPASEforiShell, sIBMPASEforiEnv, sIBMPASEforiLangid, sIBMPASEforiCntryid,
+#endif
 	sDeprecated, sUnsupported
 } ServerOpCodes;
 
@@ -549,6 +559,12 @@
 	{ "streamlocalbindunlink", sStreamLocalBindUnlink, SSHCFG_ALL },
 	{ "allowstreamlocalforwarding", sAllowStreamLocalForwarding, SSHCFG_ALL },
 	{ "fingerprinthash", sFingerprintHash, SSHCFG_GLOBAL },
+#ifdef __PASE__
+	{ "ibmpaseforishell",   sIBMPASEforiShell,   SSHCFG_GLOBAL},
+	{ "ibmpaseforienv",     sIBMPASEforiEnv,     SSHCFG_GLOBAL},
+	{ "ibmpaseforilangid",  sIBMPASEforiLangid,  SSHCFG_GLOBAL},
+	{ "ibmpaseforicntryid", sIBMPASEforiCntryid, SSHCFG_GLOBAL},
+#endif
 	{ NULL, sBadOption, 0 }
 };
 
@@ -1672,6 +1688,17 @@
 			options->adm_forced_command = xstrdup(cp + len);
 		return 0;
 
+#ifdef __PASE__
+	case sIBMPASEforiShell:
+		charptr = &options->pase_shell;
+		goto parse_filename;
+
+	/* These are ignored */
+	case sIBMPASEforiEnv:
+	case sIBMPASEforiLangid:
+	case sIBMPASEforiCntryid:
+		return 0;
+#endif
 	case sChrootDirectory:
 		charptr = &options->chroot_directory;
 
