diff -u a/configure.ac b/configure.ac
--- a/configure.ac	2015-06-30 21:35:31.000000000 -0500
+++ b/configure.ac	2017-01-10 10:14:30.085583817 -0600
@@ -566,6 +566,40 @@
 	AC_DEFINE([PTY_ZEROREAD], [1], [read(1) can return 0 for a non-closed fd])
 	AC_DEFINE([PLATFORM_SYS_DIR_UID], 2, [System dirs owned by bin (uid 2)])
 	;;
+	
+*-*-os400)
+    LIBS="$LIBS -lpthreads -liconv"
+    
+    if test "$GCC" = "yes"; then
+        CFLAGS="$CFLAGS -D__PASE__"
+    else
+        CFLAGS="$CFLAGS -D__PASE__ -qmaxmem=-1 -qalign=natural -qldbl128"
+    fi
+    
+    AC_MSG_CHECKING([how to specify blibpath for linker ($LD)])
+    if (test -z "$blibpath"); then
+        blibpath="/usr/lib:/lib"
+    fi
+    saved_LDFLAGS="$LDFLAGS"
+    if test "$GCC" = "yes"; then
+        flags="-Wl,-blibpath: -Wl,-rpath, -blibpath:"
+    else
+        flags="-blibpath: -Wl,-blibpath: -Wl,-rpath,"
+    fi
+    for tryflags in $flags ;do
+        if (test -z "$blibflags"); then
+            LDFLAGS="$saved_LDFLAGS $tryflags$blibpath"
+            AC_LINK_IFELSE([AC_LANG_PROGRAM([[]], [[]])],
+            [blibflags=$tryflags], [])
+        fi
+    done
+    if (test -z "$blibflags"); then
+        AC_MSG_RESULT([not found])
+        AC_MSG_ERROR([*** must be able to specify blibpath on PASE - check config.log])
+    else
+        AC_MSG_RESULT([$blibflags])
+    fi
+    ;;
 *-*-android*)
 	AC_DEFINE([DISABLE_UTMP], [1], [Define if you don't want to use utmp])
 	AC_DEFINE([DISABLE_WTMP], [1], [Define if you don't want to use wtmp])
Common subdirectories: a/contrib and b/contrib
