--- openssh-6.9p1/regress/Makefile	2015-06-30 21:35:31.000000000 -0500
+++ openssh-6.9p1.pase/regress/Makefile	2016-04-20 18:15:24.078627722 -0500
@@ -47,7 +47,6 @@
 		sftp-badcmds \
 		sftp-batch \
 		sftp-glob \
-		sftp-perm \
 		reconfigure \
 		dynamic-forward \
 		forwarding \
@@ -74,7 +73,9 @@
 		hostkey-agent \
 		keygen-knownhosts \
 		hostkey-rotate \
-		principals-command
+		principals-command \
+# 		sftp-perm \
+# 		
 
 
 #		dhgex \
