--- a/uidswap.c	2015-06-30 21:35:31.000000000 -0500
+++ b/uidswap.c	2016-02-08 19:28:17.562258222 -0600
@@ -49,8 +49,13 @@
 /* Saved effective uid. */
 static int	privileged = 0;
 static int	temporarily_use_uid_effective = 0;
+#ifndef __PASE__
 static gid_t	*saved_egroups = NULL, *user_groups = NULL;
 static int	saved_egroupslen = -1, user_groupslen = -1;
+#else
+static profile_handle_t prev_handle;
+static profile_handle_t curr_handle;
+#endif
 
 /*
  * Temporarily changes to the given uid.  If the effective user
@@ -59,6 +64,7 @@
 void
 temporarily_use_uid(struct passwd *pw)
 {
+#ifndef __PASE__
 	/* Save the current euid, and egroups. */
 #ifdef SAVED_IDS_WORK_WITH_SETEUID
 	saved_euid = geteuid();
@@ -129,6 +135,15 @@
 	if (seteuid(pw->pw_uid) == -1)
 		fatal("seteuid %u: %.100s", (u_int)pw->pw_uid,
 		    strerror(errno));
+#else /* __PASE__ */
+	if(pase_temporarily_use_uid(pw, &prev_handle, &curr_handle) != 0) {
+		privileged = 0;
+		return;
+	}
+	
+	privileged = 1;
+	temporarily_use_uid_effective = 1;
+#endif
 }
 
 void
@@ -170,6 +185,7 @@
 	if (!temporarily_use_uid_effective)
 		fatal("restore_uid: temporarily_use_uid not effective");
 
+#ifndef __PASE__
 #ifdef SAVED_IDS_WORK_WITH_SETEUID
 	debug("restore_uid: %u/%u", (u_int)saved_euid, (u_int)saved_egid);
 	/* Set the effective uid back to the saved privileged uid. */
@@ -190,6 +206,12 @@
 	if (setgroups(saved_egroupslen, saved_egroups) < 0)
 		fatal("setgroups: %.100s", strerror(errno));
 	temporarily_use_uid_effective = 0;
+#else /* __PASE__ */
+	if(pase_restore_uid(&prev_handle, &curr_handle) < 0) {
+		fatal("pase_restore_uid encountered error");
+	}
+	temporarily_use_uid_effective = 0;
+#endif
 }
 
 /*
@@ -227,7 +249,7 @@
 	if (setresuid(pw->pw_uid, pw->pw_uid, pw->pw_uid) < 0)
 		fatal("setresuid %u: %.100s", (u_int)pw->pw_uid, strerror(errno));
 
-#ifndef HAVE_CYGWIN
+#if !defined(HAVE_CYGWIN) && !defined(__PASE__)
 	/* Try restoration of GID if changed (test clearing of saved gid) */
 	if (old_gid != pw->pw_gid && pw->pw_uid != 0 &&
 	    (setgid(old_gid) != -1 || setegid(old_gid) != -1))
@@ -241,7 +263,7 @@
 		    (u_int)pw->pw_gid);
 	}
 
-#ifndef HAVE_CYGWIN
+#if !defined(HAVE_CYGWIN) && !defined(__PASE__)
 	/* Try restoration of UID if changed (test clearing of saved uid) */
 	if (old_uid != pw->pw_uid &&
 	    (setuid(old_uid) != -1 || seteuid(old_uid) != -1))
